include/grpcpp/impl/codegen/server_callback.h

```diff
@@ -55,18 +56,92 @@ class ServerCallbackRpcController {
   virtual void SendInitialMetadata(std::function<void(bool)>) = 0;
 };
 
+class ServerBidiReactor {
+ public:
+  virtual ~ServerBidiReactor() {}
+  virtual void OnSendInitialMetadataDone(bool ok) {}
+  virtual void OnReadDone(bool ok) {}
+  virtual void OnWriteDone(bool ok) {}
+};
+
+class ServerReadReactor {
+ public:
+  virtual ~ServerReadReactor() {}
+  virtual void OnSendInitialMetadataDone(bool ok) {}
+  virtual void OnReadDone(bool ok) {}
+};
+
+class ServerWriteReactor {
+ public:
+  virtual ~ServerWriteReactor() {}
+  virtual void OnSendInitialMetadataDone(bool ok) {}
+  virtual void OnWriteDone(bool ok) {}
+};
+
+template <class Request>
+class ServerCallbackReader {
+ public:
+  virtual ~ServerCallbackReader() {}
+
+  virtual void BindReactor(ServerReadReactor* reactor) = 0;
+  virtual void Finish(Status s) = 0;
+  virtual void SendInitialMetadata() = 0;
+  virtual void Read(Request* msg) = 0;
+};
+
+template <class Response>
+class ServerCallbackWriter {
+ public:
+  virtual ~ServerCallbackWriter() {}
+
+  virtual void BindReactor(ServerWriteReactor* reactor) = 0;
+  virtual void Finish(Status s) = 0;
+  virtual void SendInitialMetadata() = 0;
+  void Write(const Response* msg) { Write(msg, WriteOptions()); }
+  virtual void Write(const Response* msg, WriteOptions options) = 0;
+  virtual void WriteAndFinish(const Response* msg, WriteOptions options,
+                              Status s) {
+    // Default implementation that can/should be overridden
+    Write(msg, std::move(options));
+    Finish(std::move(s));
+  };
+  void WriteLast(const Response* msg, WriteOptions options) {
+    Write(msg, options.set_last_message());
+  }
+};
+
+template <class Request, class Response>
+class ServerCallbackReaderWriter {
+ public:
+  virtual ~ServerCallbackReaderWriter() {}
+
+  virtual void BindReactor(ServerBidiReactor* reactor) = 0;
+  virtual void Finish(Status s) = 0;
+  virtual void SendInitialMetadata() = 0;
+  virtual void Read(Request* msg) = 0;
+  void Write(const Response* msg) { Write(msg, WriteOptions()); }
+  virtual void Write(const Response* msg, WriteOptions options) = 0;
+  virtual void WriteAndFinish(const Response* msg, WriteOptions options,
+                              Status s) {
+    Write(msg, std::move(options));
+    Finish(std::move(s));
+  };
+  void WriteLast(const Response* msg, WriteOptions options) {
+    Write(msg, options.set_last_message());
+  }
+};
```

Comment:
Why do we need a seperate WriteLast function if all it does is set an option. The user either way needs to specify that it is writing the last message

Reply:
It's not needed, but interestingly, it is one of the few WriteOptions features that is widely used. That's why I parroted it over from the sync and CQ APIs.